folly/random/xoshiro256pp.h

```diff
@@ -138,7 +138,109 @@ class xoshiro256pp {
+      uint64x2_t s1_1 = curState1[1];
+      uint64x2_t s1_2 = curState1[2];
+      uint64x2_t s1_3 = curState1[3];
+
+      uint64x2_t s2_0 = curState2[0];
+      uint64x2_t s2_1 = curState2[1];
+      uint64x2_t s2_2 = curState2[2];
+      uint64x2_t s2_3 = curState2[3];
+
+      uint64x2_t s3_0 = curState3[0];
+      uint64x2_t s3_1 = curState3[1];
+      uint64x2_t s3_2 = curState3[2];
+      uint64x2_t s3_3 = curState3[3];
+
+      uint64x2_t sum0 = vaddq_u64(s0_0, s0_3);
+      uint64x2_t rot0 = vorrq_u64(vshlq_n_u64(sum0, 23),vshrq_n_u64(sum0, 64 - 23));
+      vecRes[i] = vaddq_u64(rot0, s0_0);
+
+      uint64x2_t sum1 = vaddq_u64(s1_0, s1_3);
+      uint64x2_t rot1 = vorrq_u64(vshlq_n_u64(sum1, 23),vshrq_n_u64(sum1, 64 - 23));
+      vecRes[i+1] = vaddq_u64(rot1, s1_0);
+
+      uint64x2_t sum2 = vaddq_u64(s2_0, s2_3);
+      uint64x2_t rot2 = vorrq_u64(vshlq_n_u64(sum2, 23),vshrq_n_u64(sum2, 64 - 23));
+      vecRes[i+2] = vaddq_u64(rot2, s2_0);
+
+      uint64x2_t sum3 = vaddq_u64(s3_0, s3_3);
+      uint64x2_t rot3 = vorrq_u64(vshlq_n_u64(sum3, 23),vshrq_n_u64(sum3, 64 - 23));
+      vecRes[i+3] = vaddq_u64(rot3, s3_0);
+
+      uint64x2_t t0 = vshlq_n_u64(s0_1, 17);
+      uint64x2_t t1 = vshlq_n_u64(s1_1, 17);
+      uint64x2_t t2 = vshlq_n_u64(s2_1, 17);
+      uint64x2_t t3 = vshlq_n_u64(s3_1, 17);
+
+      s0_2 = veorq_u64(s0_2, s0_0);
+      s0_3 = veorq_u64(s0_3, s0_1);
+      s0_1 = veorq_u64(s0_1, s0_2);
+      s0_0 = veorq_u64(s0_0, s0_3);
+      s0_2 = veorq_u64(s0_2, t0);
+      s0_3 = vorrq_u64(vshlq_n_u64(s0_3, 45), vshrq_n_u64(s0_3, 19)); // rotl(s0_3, 45)
+
+      s1_2 = veorq_u64(s1_2, s1_0);
+      s1_3 = veorq_u64(s1_3, s1_1);
+      s1_1 = veorq_u64(s1_1, s1_2);
+      s1_0 = veorq_u64(s1_0, s1_3);
+      s1_2 = veorq_u64(s1_2, t1);
+      s1_3 = vorrq_u64(vshlq_n_u64(s1_3, 45), vshrq_n_u64(s1_3, 19)); // rotl(s1_3, 45)
+
+      s2_2 = veorq_u64(s2_2, s2_0);
+      s2_3 = veorq_u64(s2_3, s2_1);
+      s2_1 = veorq_u64(s2_1, s2_2);
+      s2_0 = veorq_u64(s2_0, s2_3);
+      s2_2 = veorq_u64(s2_2, t2);
+      s2_3 = vorrq_u64(vshlq_n_u64(s2_3, 45), vshrq_n_u64(s2_3, 19)); // rotl(s2_3, 45)
+
+      s3_2 = veorq_u64(s3_2, s3_0);
+      s3_3 = veorq_u64(s3_3, s3_1);
+      s3_1 = veorq_u64(s3_1, s3_2);
+      s3_0 = veorq_u64(s3_0, s3_3);
+      s3_2 = veorq_u64(s3_2, t3);
+      s3_3 = vorrq_u64(vshlq_n_u64(s3_3, 45), vshrq_n_u64(s3_3, 19)); // rotl(s3_3, 45)
+
+      curState0[0] = s0_0;
+      curState0[1] = s0_1;
+      curState0[2] = s0_2;
+      curState0[3] = s0_3;
+
+      curState1[0] = s1_0;
+      curState1[1] = s1_1;
+      curState1[2] = s1_2;
+      curState1[3] = s1_3;
+
+      curState2[0] = s2_0;
+      curState2[1] = s2_1;
+      curState2[2] = s2_2;
+      curState2[3] = s2_3;
+
+      curState3[0] = s3_0;
+      curState3[1] = s3_1;
+      curState3[2] = s3_2;
+      curState3[3] = s3_3;
+
+    }
+    cur = 0;
+#else
+   for (uint64_t i = 0; i < VecResCount; i++) {
```

Comment:
What code gen do you get if you add #pragma clang loop unroll_count(4) to unroll this loop 4 times? Is it still worse than the manually unrolled loop above?

Reply:
Would it make sense to set the loop unroll count to the number of SIMD units?